From 3128f6c4c058638cae715c219fed416ac5605c1e Mon Sep 17 00:00:00 2001
From: Jiali Chen <chenjiali@radxa.com>
Date: Fri, 6 Feb 2026 09:05:23 +0000
Subject: [PATCH 055/268] drm/tiny/ili9486: add support for
 waveshare,rpi-lcd-35-b

Add support for the Waveshare 3.5inch RPi LCD (B) to the ili9486
driver. This variant requires different initialization sequences:
1. Inverted color mode enabled via command 0x21
2. Different gamma correction parameters for PGAMCTRL and NGAMCTRL

The initialization sequence is from:
https://github.com/radxa-pkg/radxa-overlays/blob/569043b975d3f659e561e25539a5d854192adf26/arch/arm64/boot/dts/rockchip/overlays/rk3568-spi3-m1-cs0-waveshare35b.dts#L12-L25

Changes include:
- Add "waveshare,rpi-lcd-35-b" to the device tree compatible list
- Add corresponding SPI device ID entry
- Modify the waveshare_enable() function to detect the B variant and
  apply appropriate settings:
  * Send command 0x21 to enable inverted color mode
  * Use different PGAMCTRL and NGAMCTRL gamma values
  * Keep DGAMCTRL unchanged for both variants

This allows the driver to properly initialize both the original
waveshare,rpi-lcd-35 and the new waveshare,rpi-lcd-35-b displays.

Signed-off-by: Jiali Chen <chenjiali@radxa.com>
---
 drivers/gpu/drm/tiny/ili9486.c | 26 ++++++++++++++++++++------
 1 file changed, 20 insertions(+), 6 deletions(-)

--- a/drivers/gpu/drm/tiny/ili9486.c
+++ b/drivers/gpu/drm/tiny/ili9486.c
@@ -101,6 +101,8 @@ static void waveshare_enable(struct drm_
 	struct mipi_dbi *dbi = &dbidev->dbi;
 	u8 addr_mode;
 	int ret, idx;
+	bool is_rpi_lcd_35_b = device_is_compatible(dbidev->drm.dev,
+						    "waveshare,rpi-lcd-35-b");
 
 	if (!drm_dev_enter(pipe->crtc.dev, &idx))
 		return;
@@ -123,12 +125,22 @@ static void waveshare_enable(struct drm_
 
 	mipi_dbi_command(dbi, ILI9486_VMCTRL1, 0x00, 0x00, 0x00, 0x00);
 
-	mipi_dbi_command(dbi, ILI9486_PGAMCTRL,
-			 0x0F, 0x1F, 0x1C, 0x0C, 0x0F, 0x08, 0x48, 0x98,
-			 0x37, 0x0A, 0x13, 0x04, 0x11, 0x0D, 0x0);
-	mipi_dbi_command(dbi, ILI9486_NGAMCTRL,
-			 0x0F, 0x32, 0x2E, 0x0B, 0x0D, 0x05, 0x47, 0x75,
-			 0x37, 0x06, 0x10, 0x03, 0x24, 0x20, 0x00);
+	if (is_rpi_lcd_35_b) {
+		mipi_dbi_command(dbi, 0x21);
+		mipi_dbi_command(dbi, ILI9486_PGAMCTRL,
+				 0x00, 0x13, 0x18, 0x04, 0x0F, 0x06, 0x3A, 0x56,
+				 0x4D, 0x03, 0x0A, 0x06, 0x30, 0x3E, 0x0F);
+		mipi_dbi_command(dbi, ILI9486_NGAMCTRL,
+				 0x00, 0x13, 0x18, 0x01, 0x11, 0x06, 0x38, 0x34,
+				 0x4D, 0x06, 0x0D, 0x0B, 0x31, 0x37, 0x0F);
+	} else {
+		mipi_dbi_command(dbi, ILI9486_PGAMCTRL,
+				 0x0F, 0x1F, 0x1C, 0x0C, 0x0F, 0x08, 0x48, 0x98,
+				 0x37, 0x0A, 0x13, 0x04, 0x11, 0x0D, 0x0);
+		mipi_dbi_command(dbi, ILI9486_NGAMCTRL,
+				 0x0F, 0x32, 0x2E, 0x0B, 0x0D, 0x05, 0x47, 0x75,
+				 0x37, 0x06, 0x10, 0x03, 0x24, 0x20, 0x00);
+	}
 	mipi_dbi_command(dbi, ILI9486_DGAMCTRL,
 			 0x0F, 0x32, 0x2E, 0x0B, 0x0D, 0x05, 0x47, 0x75,
 			 0x37, 0x06, 0x10, 0x03, 0x24, 0x20, 0x00);
@@ -183,6 +195,7 @@ static const struct drm_driver ili9486_d
 
 static const struct of_device_id ili9486_of_match[] = {
 	{ .compatible = "waveshare,rpi-lcd-35" },
+	{ .compatible = "waveshare,rpi-lcd-35-b" },
 	{ .compatible = "ozzmaker,piscreen" },
 	{},
 };
@@ -191,6 +204,7 @@ MODULE_DEVICE_TABLE(of, ili9486_of_match
 static const struct spi_device_id ili9486_id[] = {
 	{ "ili9486", 0 },
 	{ "rpi-lcd-35", 0 },
+	{ "rpi-lcd-35-b", 0 },
 	{ "piscreen", 0 },
 	{ }
 };
